src/main/java/com/billyyccc/mssqlclient/impl/codec/InitCommandCodec.java

```diff
@@ -0,0 +1,224 @@
+
+    // AtchDBFile
+    offsetInfo.put("AtchDBFile", packet.writerIndex());
+    packet.writeShortLE(0x00); // offset
+    packet.writeShortLE(0x00);
+
+    // ChangePassword
+    offsetInfo.put("ChangePassword", packet.writerIndex());
+    packet.writeShortLE(0x00); // offset
+    packet.writeShortLE(0x00);
+
+    // SSPILong
+    packet.writeIntLE(0x00);
+
+      /*
+        Data part: note we should set offset by calculation before writing data
+       */
+    packet.setShortLE(offsetInfo.get("HostName"), packet.writerIndex() - startIdx);
+    packet.writeCharSequence(hostName, CharsetUtil.UTF_16LE);
+
+    packet.setShortLE(offsetInfo.get("UserName"), packet.writerIndex() - startIdx);
+    packet.writeCharSequence(userName, CharsetUtil.UTF_16LE);
+
+    packet.setShortLE(offsetInfo.get("Password"), packet.writerIndex() - startIdx);
+    writePassword(packet, password);
+
+    packet.setShortLE(offsetInfo.get("AppName"), packet.writerIndex() - startIdx);
+    packet.writeCharSequence(appName, CharsetUtil.UTF_16LE);
+
+    packet.setShortLE(offsetInfo.get("ServerName"), packet.writerIndex() - startIdx);
+    packet.writeCharSequence(serverName, CharsetUtil.UTF_16LE);
+
+    packet.setShortLE(offsetInfo.get("Unused"), packet.writerIndex() - startIdx);
+
+    packet.setShortLE(offsetInfo.get("CltIntName"), packet.writerIndex() - startIdx);
+    packet.writeCharSequence(interfaceLibraryName, CharsetUtil.UTF_16LE);
+
+    packet.setShortLE(offsetInfo.get("Language"), packet.writerIndex() - startIdx);
+
+    packet.setShortLE(offsetInfo.get("Database"), packet.writerIndex() - startIdx);
+    packet.writeCharSequence(database, CharsetUtil.UTF_16LE);
+
+    packet.setShortLE(offsetInfo.get("SSPI"), packet.writerIndex() - startIdx);
+
+    packet.setShortLE(offsetInfo.get("AtchDBFile"), packet.writerIndex() - startIdx);
+
+    packet.setShortLE(offsetInfo.get("ChangePassword"), packet.writerIndex() - startIdx);
+
+    // set length
+    packet.setIntLE(startIdx, packet.writerIndex() - startIdx);
+
+    int packetLen = packet.writerIndex() - startIdx + 8;
+    packet.setShort(packetLenIdx, packetLen);
+
+    chctx.writeAndFlush(packet);
+
+  }
+
+  /*
+    Before submitting a password from the client to the server,
+    for every byte in the password buffer starting with the position pointed to by ibPassword or ibChangePassword,
+    the client SHOULD first swap the four high bits with the four low bits and then do a bit-XOR with 0xA5 (10100101).
+    After reading a submitted password, for every byte in the password buffer starting with the position pointed to by ibPassword or ibChangePassword,
+    the server SHOULD first do a bit-XOR with 0xA5 (10100101) and then swap the four high bits with the four low bits.
+   */
+  private void writePassword(ByteBuf payload, String password) {
+    byte[] bytes = password.getBytes(CharsetUtil.UTF_16LE);
+    for (byte b : bytes) {
+      payload.writeByte((b >> 4 | ((b & 0x0F) << 4)) ^ 0xA5);
```

Comment:
do instead this transformation on the `bytes` array directly then use `ByteBuf#writeBytes` passing the array, this will be more efficient